### PR TITLE
Codec-related refactorings and quality of life improvements

### PR DESCRIPTION
**Pull request by neysofu**
This (draft, at the time of writing) PR makes the following changes in an attempt to improve code quality in areas of the codebase that are related to `svm-codec`:

- Reduce macro usage; implement a trait `BytesPrimitive: AsRef<[u8]> + AsMut<[u8]>` for `Address`, `TemplateAddr`, `State` and others.
- Leverage the new `BytesPrimitive` trait to implement common features with generics, as well as unifying encoding and decoding logic.
- Unify `encode_xyz` and `decode_xyz` functions in `svm_codec` under a `Codec` trait. This helps discoverability, can benefit from writing generic unit tests and improves error handling.
- Overhaul the `wasm` API. Rewrite buffering logic with safety in mind and greatly reduce amount of code.
- Exploratory work for providing a Typescript typed wrapper around `svm_codec::api::wasm`.

### CODE REVIEW

File: crates/codec/src/api/wasm/buffer.rs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        let mut buf = Self::alloc(len);
+
+        buf.as_mut()[0] = Self::OK_MARKER;
+        buf.as_mut()[1..].clone_from_slice(data);
+
+        buf
+    }
+
+    #[allow(unused)]
+    pub fn as_result(&self) -> Option<Result<&[u8], &str>> {
```
Comment (YaronWittenstein) [inline crates/codec/src/api/wasm/buffer.rs:96]: Do really need that function? (also, name feels a bit confusing to me)

File: crates/codec/src/api/wasm/buffer.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        let mut buf = Self(Some(vec), true);
+
+        buf.set_capacity(len);
+        buf.set_len(len);
+
+        buf
+    }
+
+    #[cfg(test)]
+    pub fn then_free(mut self) -> Self {
```
Comment (YaronWittenstein) [inline crates/codec/src/api/wasm/buffer.rs:80]: Why not name it `free`?

File: crates/codec/src/api/wasm/buffer.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+    pub fn len(&self) -> u32 {
+        u32::from_be_bytes((&self.vec()[Self::LEN]).try_into().unwrap())
+    }
+
+    pub fn capacity(&self) -> u32 {
+        u32::from_be_bytes((&self.vec()[Self::CAPACITY]).try_into().unwrap())
+    }
+
+    pub fn set_len(&mut self, len: u32) {
```
Comment (YaronWittenstein) [inline crates/codec/src/api/wasm/buffer.rs:163]: I don't think we need `set_len` to be public. The current implementation just allocates a buffer once.
Why can't we just have the `alloc_ok / alloc_err` above?

File: crates/codec/src/api/wasm/buffer.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        assert!(len <= self.capacity());
+
+        self.vec_mut()[Self::LEN].clone_from_slice(&len.to_be_bytes());
+    }
+
+    pub fn set_capacity(&mut self, capacity: u32) {
```
Comment (YaronWittenstein) [inline crates/codec/src/api/wasm/buffer.rs:169]: Same. I don't think we really need a public `set_capacity`

File: crates/codec/src/ext.rs [hunk 5]
```diff
@@ -1,42 +1,30 @@
...
-    fn read_bool(&mut self) -> Result<bool>;
-
-    /// Reads an unsigned 16-bit integer (Big-Endian)
-    fn read_u16_be(&mut self) -> Result<u16>;
-
-    /// Reads an unsigned 32-bit integer (Big-Endian)
-    fn read_u32_be(&mut self) -> Result<u32>;
+pub trait ReadExt: Sized {
+    /// Tries to read the next byte, but doesn't move the cursor forward.
+    fn seek_byte(&self) -> Option<u8>;
```
Comment (YaronWittenstein) [inline crates/codec/src/ext.rs:8]: I think that `peek_byte` would make a better term here

File: crates/runtime/src/env/default/serialize.rs [hunk 6]
```diff
@@ -23,46 +17,21 @@ impl AccountSerializer for DefaultAccountSerializer {
...
-
-fn encode_spawner(account: &ExtAccount, w: &mut Vec<u8>) {
-    let spawner = account.spawner();
-    w.write_address(spawner);
-}
-
 impl AccountDeserializer for DefaultAccountDeserializer {
     fn deserialize(bytes: &[u8]) -> Option<ExtAccount> {
-        let mut cursor = Cursor::new(bytes);
+        let mut reader = Cursor::new(bytes);
```
Comment (YaronWittenstein) [inline crates/runtime/src/env/default/serialize.rs:30]: I would keep the name `cursor` here (since the type is `Cursor` as well)

File: crates/runtime/tests/runtime_tests.rs [hunk 7]
```diff
@@ -136,27 +136,25 @@ fn memory_runtime_validate_deploy_ok() {
...
     let message = vec![0xFF, 0xFF];
 
-    let error = ParseError::NotEnoughBytes(Field::Address);
-    let expected = ValidateError::Parse(error);
-
-    let actual = runtime.validate_spawn(&message).unwrap_err();
-    assert_eq!(expected, actual);
+    assert!(matches!(
+        runtime.validate_spawn(&message),
+        Err(ValidateError::Parse(ParseError::Eof))
```
Comment (YaronWittenstein) [inline crates/runtime/tests/runtime_tests.rs:145]: The previous version of the test had a different error, it said what field was missing bytes

File: crates/runtime/tests/runtime_tests.rs [hunk 8]
```diff
@@ -3,24 +3,24 @@ use svm_sdk as sdk;
...
+fn memory_runtime_validate_deploy_eof() {
     let runtime = testing::create_memory_runtime();
     let message = vec![0xFF, 0xFF];
 
-    let error = ParseError::NotEnoughBytes(Field::SectionKind);
-    let expected = ValidateError::Parse(error);
-
-    let actual = runtime.validate_deploy(&message).unwrap_err();
-    assert_eq!(expected, actual);
+    assert!(matches!(
```
Comment (YaronWittenstein) [inline crates/runtime/tests/runtime_tests.rs:20]: The previous version of the test had a different error, it said what field was missing bytes

File: crates/runtime/tests/runtime_tests.rs [hunk 9]
```diff
@@ -136,27 +136,25 @@ fn memory_runtime_validate_deploy_ok() {
...
+fn memory_runtime_validate_call_eof() {
     let runtime = testing::create_memory_runtime();
     let message = vec![0xFF, 0xFF];
 
-    let error = ParseError::NotEnoughBytes(Field::TargetAddr);
-    let expected = Err(ValidateError::Parse(error));
-
-    let actual = runtime.validate_call(&message);
-    assert_eq!(expected, actual);
+    assert!(matches!(
```
Comment (YaronWittenstein) [inline crates/runtime/tests/runtime_tests.rs:154]: The previous version of the test had a different error, it said what field was missing bytes

File: crates/runtime/tests/vmcalls_tests.rs [hunk 10]
```diff
@@ -247,7 +247,7 @@ fn vmcalls_load160() {
     let view = &memory.view::<u8>()[ptr as usize..(ptr as usize + 20)];
     let bytes: Vec<u8> = view.iter().map(|cell| cell.get()).collect();
 
-    assert_eq!(target_addr, Address::from(&bytes[..]));
+    assert_eq!(target_addr, Address::new(&bytes[..]));
```
Comment (YaronWittenstein) [inline crates/runtime/tests/vmcalls_tests.rs:250]: Why not keep the previous `from` ?

File: crates/types/src/lib.rs [hunk 11]
```diff
@@ -40,6 +37,110 @@ pub use receipt::{
...
+        Self::from(buf)
+    }
+
+    /// Returns a reference to the raw contents of `self`.
+    fn as_slice(&self) -> &[u8] {
+        self.as_ref()
+    }
+
+    /// Creates a new `Self` with all bytes set to zero.
+    fn zeros() -> Self {
```
Comment (YaronWittenstein) [inline crates/types/src/lib.rs:88]: can we turn it into `const fn` ?

File: crates/types/src/lib.rs [hunk 12]
```diff
@@ -40,6 +37,110 @@ pub use receipt::{
...
+        [0; N].into()
+    }
+
+    /// Checks whether or not `self` is equal to [`BytesPrimitive::zeros()`].
+    fn is_zeros(&self) -> bool {
+        Self::zeros().as_ref() == self.as_ref()
+    }
+
+    /// Fills a new instance of [`Self`] wihh `byte`.
```
Comment (YaronWittenstein) [inline crates/types/src/lib.rs:97]: typo: `wihh`

File: crates/types/src/lib.rs [hunk 13]
```diff
@@ -40,6 +37,110 @@ pub use receipt::{
...
+    fn repeat(byte: u8) -> Self {
+        [byte; N].into()
+    }
+
+    /// Returns the constant number of bytes that make up instances of `Self`.
+    fn len() -> usize {
```
Comment (YaronWittenstein) [inline crates/types/src/lib.rs:103]: can we turn it into `const fn`?

File: crates/codec/src/api/json/deploy.rs [hunk 14]
```diff
@@ -40,11 +49,7 @@ fn to_data_layout(blob: Vec<u8>) -> Result<Layout, JsonError> {
 
     let data: Vec<u32> = blob
         .chunks_exact(4)
-        .map(|buf| {
-            let bytes: [u8; 4] = [buf[0], buf[1], buf[2], buf[3]];
-
-            u32::from_be_bytes(bytes)
-        })
+        .map(|bytes| u32::from_be_bytes(bytes.try_into().unwrap()))
```
Comment (YaronWittenstein) [inline crates/codec/src/api/json/deploy.rs:52]: nice!

File: crates/codec/src/api/json/mod.rs [hunk 15]
```diff
@@ -66,3 +54,25 @@ pub(crate) fn logs_to_json(logs: &[ReceiptLog]) -> Vec<Json> {
         })
         .collect()
 }
+
+pub(crate) fn get_field<T>(json: &mut Json, name: &str) -> Result<T, JsonError>
```
Comment (YaronWittenstein) [inline crates/codec/src/api/json/mod.rs:58]: nice!

**Review by YaronWittenstein — CHANGES_REQUESTED** [review]
The PR looks good!
(a bit hard to follow all the details)

File: crates/codec/src/api/wasm/buffer.rs [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+    const HEADER_SIZE: usize = 8;
+
+    const OK_MARKER: u8 = 1;
+    const ERR_MARKER: u8 = 0;
+
+    pub fn alloc(len: u32) -> Self {
+        let mut vec = vec![0; len as usize + Self::HEADER_SIZE];
+        vec.shrink_to_fit();
+
+        assert_eq!(vec.capacity(), len as usize + Self::HEADER_SIZE);
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs:68]: :+1:

File: crates/codec/src/api/wasm/buffer.rs [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+        let mut buf = Self::alloc(len);
+
+        buf.as_mut()[0] = Self::ERR_MARKER;
+        buf.as_mut()[1..].clone_from_slice(string.as_bytes());
+
+        buf
+    }
+
+    pub fn free(mut self) {
+        self.0.take();
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs]: Done!

File: crates/codec/src/api/wasm/buffer.rs [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+/// Contains the raw data of the buffer.
+///
+/// Allocates a new WASM buffer having `Data` of `length` bytes.
+///
+/// The total allocation size of the buffer will always be bigger due to the
+/// `Header` section.
+/// If for the `capacity` of the `Data` will be bigger - it will also increase
+/// the amount of allocated data.
+#[derive(Debug, PartialEq)]
+pub struct Buffer(Option<Vec<u8>>, bool);
```
[the remark this answers is not in the document]
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs]: Good points, will do now.

File: crates/codec/src/api/wasm/buffer.rs (hunk 1 above)
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs:111]: You're right, I thought it would later be useful for tests but it's actually unused.

File: crates/codec/src/api/wasm/buffer.rs (hunk 2 above)
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs:94]: Because it doesn't free `Buffer`. `Buffer` has two modes - either leaking (default, given that it's what you usually want when doing FFI) or freeing. This changes the memory mode, but it doesn't immediately free. Memory will only get freed on drop.

File: crates/codec/src/api/wasm/buffer.rs (hunk 3 above)
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs]: Yup, I agree!

File: crates/runtime/tests/vmcalls_tests.rs (hunk 10 above)
Comment (neysofu) [reply crates/runtime/tests/vmcalls_tests.rs:250]: We can't implement [`std::convert::From` for `BytesPrimitive` implementors, we can only provide `BytesPrimitive::foobar` methods. A rename seemed like a small change in exchange for a number of benefits.

File: crates/types/src/lib.rs (hunk 11 above)
Comment (neysofu) [reply crates/types/src/lib.rs:88]: We will in the future, for now Rust doesn't support `const fn` in traits.

File: crates/types/src/lib.rs (hunk 12 above)
Comment (neysofu) [reply crates/types/src/lib.rs]: Fixed.

File: crates/types/src/lib.rs (hunk 13 above)
Comment (neysofu) [reply crates/types/src/lib.rs:103]: Not yet, Rust doesn't support `const fn` in traits. We will be able to do this in the future though.

File: crates/codec/src/api/wasm/buffer.rs (hunk 4 above)
Comment (neysofu) [reply crates/codec/src/api/wasm/buffer.rs]: Removed `pub` :+1:

File: crates/codec/src/ext.rs (hunk 5 above)
Comment (neysofu) [reply crates/codec/src/ext.rs]: Renamed it.

File: crates/runtime/src/env/default/serialize.rs (hunk 6 above)
Comment (neysofu) [reply crates/runtime/src/env/default/serialize.rs]: :+1:

File: crates/runtime/tests/runtime_tests.rs (hunk 8 above)
Comment (neysofu) [reply crates/runtime/tests/runtime_tests.rs:20]: As agreed in chat, this is good enough considering that `Codec::decode` only deals with auto-generated inputs.

File: crates/runtime/tests/runtime_tests.rs (hunk 9 above)
Comment (neysofu) [reply crates/runtime/tests/runtime_tests.rs:154]: As above.

**Comment by neysofu** [comment]
I've turned `BytesPrimitive::len()` into an associated constant `BytesPrimitive::N` and also fixed all other issues, I think. I'll wait for CI.
